### DISCUSSION TITLE
raw_user_meta_data field is not updated

### DISCUSSION THREAD

**Discussion by annaisdevk** [supabase/supabase#22158] [Questions] [answered]
I have a trigger function that fires `AFTER INSERT ON auth.users`.

This function performs several actions on other tables. After performing these actions, the `raw_app_meta_data` and `raw_user_meta_data` fields of `auth.users` need to be updated. Everything works correctly except the `raw_user_meta_data` update ⚠. It does not make any changes to this field. However, when I run the same query from SQL Editor, it does update.

Previously, a few months or more ago, I had used the same code and it worked. Are there any restrictions on this field? I'm looking in the documentation, but I don't see anything about this.

```
CREATE OR REPLACE FUNCTION lorem_ipsum()
RETURNS TRIGGER
LANGUAGE plpgsql
security definer set search_path = public
AS $$
BEGIN 

   --- another code
  

    --- This does not work:
    UPDATE auth.users SET raw_user_meta_data = null WHERE id = new_user_id; 

    RETURN NEW;

END;
$$;
```

```
CREATE TRIGGER lorem_ipsum_trigger
AFTER INSERT ON auth.users
FOR EACH ROW
EXECUTE FUNCTION lorem_ipsum();
```

**Comment by GaryAustin1** [comment, marked as answer]
Pretty sure gotrue/auth now does a follow up update on auth.users to set the metadata.  This means you need another trigger on auth.users update for that column changing.  This was a change made at some point.

**Reply by annaisdevk** [reply]
Great. Thank you! Now it works.

I put the code inside a function that fires in `AFTER UPDATE ON auth.users` 

For signup/signin, I use email provider with "confirm email". And since the user account is created in auth.users, the update is called several times (from the supabase code), I checked which field is the last to be updated. And I see that it is `confirmation_sent_at`. Otherwise, if I do not update it after the last field, it will overwrite my changes. 

This is what finally works for me:

```
CREATE OR REPLACE FUNCTION clear_custom_user_meta_data()
RETURNS TRIGGER
LANGUAGE plpgsql
security definer set search_path = public
AS $$
BEGIN 
  IF (NEW.confirmation_sent_at IS NOT NULL AND OLD.confirmation_sent_at IS NULL) THEN   
    UPDATE auth.users
    SET
      raw_user_meta_data = jsonb_set(raw_user_meta_data, '{custom_field_1}', 'null'::jsonb) - 'custom_field_2' - 'custom_field_3' - 'custom_field_4'
    WHERE id = NEW.id;
  END IF;

  RETURN NEW;
END;
$$;
```

```
CREATE TRIGGER clear_custom_user_meta_data_trigger
AFTER UPDATE ON auth.users
FOR EACH ROW
EXECUTE FUNCTION clear_custom_user_meta_data();
```
